dynosaur/tests/pass/basic-mut.stdout

```diff
@@ -27,56 +27,43 @@ mod _dynosaur_macro_DynMyTrait {
+                    ()> + '_>> = self.ptr.foo();
             let fut:
                     ::core::pin::Pin<Box<dyn ::core::future::Future<Output =
                     ()> + 'static>> = unsafe { ::core::mem::transmute(fut) };
             fut
         }
     }
     impl<'dynosaur_struct> DynMyTrait<'dynosaur_struct> {
-        pub fn new<DYNOSAUR>(value: DYNOSAUR) -> Self where
-            DYNOSAUR: MyTrait + 'dynosaur_struct {
-            let value = Box::new(value);
-            Self {
-                ptr: Box::into_raw(value as
-                            Box<dyn ErasedMyTrait + 'dynosaur_struct>) as
-                    *mut (dyn ErasedMyTrait + 'dynosaur_struct),
-                owned: true,
-            }
+        pub fn new(value: Box<impl MyTrait + 'dynosaur_struct>)
+            -> Box<DynMyTrait<'dynosaur_struct>> {
+            let value: Box<dyn ErasedMyTrait + 'dynosaur_struct> = value;
+            unsafe { ::core::mem::transmute(value) }
         }
-        pub fn from_ref<DYNOSAUR>(value: &'dynosaur_struct DYNOSAUR)
-            -> ::dynosaur::macro_lib::Ref<Self> where DYNOSAUR: MyTrait +
-            'dynosaur_struct {
-            let this =
-                Self {
-                    ptr: value as
-                                &'dynosaur_struct (dyn ErasedMyTrait + 'dynosaur_struct) as
-                            *const (dyn ErasedMyTrait + 'dynosaur_struct) as
-                        *mut (dyn ErasedMyTrait + 'dynosaur_struct),
-                    owned: false,
-                };
-            unsafe { ::dynosaur::macro_lib::Ref::new(this) }
+        pub fn boxed(value: impl MyTrait + 'dynosaur_struct)
+            -> Box<DynMyTrait<'dynosaur_struct>> {
+            Self::new(Box::new(value))
         }
-        pub fn from_mut<DYNOSAUR>(value: &'dynosaur_struct mut DYNOSAUR)
-            -> ::dynosaur::macro_lib::RefMut<Self> where DYNOSAUR: MyTrait +
-            'dynosaur_struct {
-            let this =
-                Self {
-                    ptr: value as
-                            &'dynosaur_struct mut (dyn ErasedMyTrait + 'dynosaur_struct)
-                        as *mut (dyn ErasedMyTrait + 'dynosaur_struct),
-                    owned: false,
-                };
-            unsafe { ::dynosaur::macro_lib::RefMut::new(this) }
+        pub fn from_ref(value:
+                &'dynosaur_struct (impl MyTrait + 'dynosaur_struct))
+            -> &'dynosaur_struct DynMyTrait<'dynosaur_struct> {
+            let value: &(dyn ErasedMyTrait + 'dynosaur_struct) = &*value;
+            unsafe { ::core::mem::transmute(value) }
+        }
+        pub fn from_mut(value:
+                &'dynosaur_struct mut (impl MyTrait + 'dynosaur_struct))
+            -> &'dynosaur_struct mut DynMyTrait<'dynosaur_struct> {
+            let value: &mut (dyn ErasedMyTrait + 'dynosaur_struct) =
+                &mut *value;
+            unsafe { ::core::mem::transmute(value) }
```

Comment:
Thinking about it more, I think we _could_ add a lifetime param to these functions to allow the self lifetime to vary from the reference lifetime. It doesn't seem likely that it would come up often but it would be annoying if you needed it and it wasn't supported.

The basic use case I was thinking of is "what if you wanted to swap two `DynMyTrait<'static>`s", but that won't work because DynMyTrait is unsized. I still feel like it might be useful in some real world cases though.

Reply:
```suggestion
        pub fn from_ref<'dynosaur_ref>(value:
                &'dynosaur_ref (impl MyTrait + 'dynosaur_struct))
            -> &'dynosaur_ref DynMyTrait<'dynosaur_struct> {
            let value: &(dyn ErasedMyTrait + 'dynosaur_struct) = &*value;
            unsafe { ::core::mem::transmute(value) }
        }
        pub fn from_mut<'dynosaur_ref>(value:
                &'dynosaur_ref mut (impl MyTrait + 'dynosaur_struct))
            -> &'dynosaur_ref mut DynMyTrait<'dynosaur_struct> {
            let value: &mut (dyn ErasedMyTrait + 'dynosaur_struct) =
                &mut *value;
            unsafe { ::core::mem::transmute(value) }
        }
```

or equivalently

```suggestion
        pub fn from_ref(value:
                &(impl MyTrait + 'dynosaur_struct))
            -> &DynMyTrait<'dynosaur_struct> {
            let value: &(dyn ErasedMyTrait + 'dynosaur_struct) = &*value;
            unsafe { ::core::mem::transmute(value) }
        }
        pub fn from_mut(value:
                &mut (impl MyTrait + 'dynosaur_struct))
            -> &mut DynMyTrait<'dynosaur_struct> {
            let value: &mut (dyn ErasedMyTrait + 'dynosaur_struct) =
                &mut *value;
            unsafe { ::core::mem::transmute(value) }
        }
```